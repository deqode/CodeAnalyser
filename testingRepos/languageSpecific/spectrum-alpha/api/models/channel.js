// @flow
const { db } = require('shared/db');
import { sendChannelNotificationQueue } from 'shared/bull/queues';
import type { DBChannel } from 'shared/types';

// reusable query parts -- begin
const channelsByCommunitiesQuery = (...communityIds: string[]) =>
  db
    .table('channels')
    .getAll(...communityIds, { index: 'communityId' })
    .filter(channel => channel.hasFields('deletedAt').not());

const channelsByIdsQuery = (...channelIds: string[]) =>
  db
    .table('channels')
    .getAll(...channelIds)
    .filter(channel => channel.hasFields('deletedAt').not());

const threadsByChannelsQuery = (...channelIds: string[]) =>
  channelsByIdsQuery(...channelIds)
    .eqJoin('id', db.table('threads'), { index: 'channelId' })
    .map(row => row('right'))
    .filter(thread => db.not(thread.hasFields('deletedAt')));

const membersByChannelsQuery = (...channelIds: string[]) =>
  channelsByIdsQuery(...channelIds)
    .eqJoin('id', db.table('usersChannels'), { index: 'channelId' })
    .map(row => row('right'))
    .filter({ isBlocked: false, isPending: false, isMember: true });

// reusable query parts -- end

// prettier-ignore
const getChannelsByCommunity = (communityId: string): Promise<Array<DBChannel>> => {
  return channelsByCommunitiesQuery(communityId).run();
};

/*
  If a non-user is viewing a community page, they should only see threads
  from public channels. We use this function to return an array of channelIds
  that are public, and pass them into a getThreads function
*/
// prettier-ignore
const getPublicChannelsByCommunity = (communityId: string): Promise<Array<string>> => {
  return channelsByCommunitiesQuery(communityId)
    .filter({ isPrivate: false })
    .filter(row => row.hasFields('archivedAt').not())
    .map(c => c('id'))
    .run();
};

/*
  If a user is viewing a community, they should see threads from all public channels as well as from private channels they are a member of.

  This function returns an array of objects with the field 'id' that corresponds
  to a channelId. This array of IDs will be passed into a threads method which
  will only return threads in those channels
*/
// prettier-ignore
const getChannelsByUserAndCommunity = async (communityId: string, userId: string): Promise<Array<string>> => {
  const channels = await channelsByCommunitiesQuery(communityId).run();
  const unarchived = channels.filter(channel => !channel.archivedAt)
  const channelIds = unarchived.map(channel => channel.id)
  
  return db
    .table('usersChannels')
    .getAll(...channelIds.map(id => ([userId, id])), {
      index: 'userIdAndChannelId',
    })
    .filter({ isMember: true })('channelId')
    .run();
};

const getChannelsByUser = (userId: string): Promise<Array<DBChannel>> => {
  return db
    .table('usersChannels')
    .getAll([userId, 'member'], [userId, 'moderator'], [userId, 'owner'], {
      index: 'userIdAndRole',
    })
    .eqJoin('channelId', db.table('channels'))
    .without({ left: ['id', 'channelId', 'userId', 'createdAt'] })
    .zip()
    .filter(channel => db.not(channel.hasFields('deletedAt')))
    .run();
};

const getChannelBySlug = async (
  channelSlug: string,
  communitySlug: string
): Promise<?DBChannel> => {
  const [communityId] = await db
    .table('communities')
    .getAll(communitySlug, { index: 'slug' })('id')
    .run();

  if (!communityId) return null;

  return db
    .table('channels')
    .getAll(communityId, { index: 'communityId' })
    .filter(channel =>
      channel('slug')
        .eq(channelSlug)
        .and(db.not(channel.hasFields('deletedAt')))
    )
    .run()
    .then(res => {
      if (Array.isArray(res) && res.length > 0) return res[0];
      return null;
    });
};

const getChannelById = async (id: string) => {
  return (await channelsByIdsQuery(id).run())[0] || null;
};

type GetChannelByIdArgs = {|
  id: string,
|};

type GetChannelBySlugArgs = {|
  slug: string,
  communitySlug: string,
|};

export type GetChannelArgs = GetChannelByIdArgs | GetChannelBySlugArgs;

const getChannels = (channelIds: Array<string>): Promise<Array<DBChannel>> => {
  return channelsByIdsQuery(...channelIds).run();
};

type GroupedCount = {
  group: string,
  reduction: number,
};

// prettier-ignore
const getChannelsThreadCounts = (channelIds: Array<string>): Promise<Array<GroupedCount>> => {
  return threadsByChannelsQuery(...channelIds)
    .group('channelId')
    .count()
    .run();
};

// prettier-ignore
const getChannelsMemberCounts = (channelIds: Array<string>): Promise<Array<GroupedCount>> => {
  return membersByChannelsQuery(...channelIds)
    .group('channelId')
    .count()
    .run();
};

export type CreateChannelInput = {
  input: {
    communityId: string,
    name: string,
    description: string,
    slug: string,
    isPrivate: boolean,
    isDefault: boolean,
  },
};

export type EditChannelInput = {
  input: {
    channelId: string,
    name: string,
    description: string,
    slug: string,
    isPrivate: Boolean,
  },
};

// prettier-ignore
const createChannel = ({ input }: CreateChannelInput, userId: string): Promise<DBChannel> => {
  const { communityId, name, slug, description, isPrivate, isDefault } = input;

  return db
    .table('channels')
    .insert(
      {
        communityId,
        createdAt: new Date(),
        name,
        description,
        slug,
        isPrivate,
        isDefault: isDefault ? true : false,
        memberCount: 0,
      },
      { returnChanges: true }
    )
    .run()
    .then(result => result.changes[0].new_val)
    .then(channel => {
      // only trigger a new channel notification is the channel is public
      if (!channel.isPrivate) {
        sendChannelNotificationQueue.add({ channel, userId });
      }

      return channel;
    });
};

// prettier-ignore
const createGeneralChannel = (communityId: string, userId: string): Promise<DBChannel> => {
  return createChannel(
    {
      input: {
        name: 'General',
        slug: 'general',
        description: 'General Chatter',
        communityId,
        isPrivate: false,
        isDefault: true,
      },
    },
    userId
  );
};

// prettier-ignore
const editChannel = async ({ input }: EditChannelInput): Promise<DBChannel> => {
  const { name, slug, description, isPrivate, channelId } = input;

  const channelRecord = await db
    .table('channels')
    .get(channelId)
    .run()
    .then(result => {
      return Object.assign({}, result, {
        name,
        description,
        slug,
        isPrivate,
      });
    });

  return db
    .table('channels')
    .get(channelId)
    .update({ ...channelRecord }, { returnChanges: 'always' })
    .run()
    .then(result => {
      // if an update happened
      if (result.replaced === 1) {
        return result.changes[0].new_val;
      }

      // an update was triggered from the client, but no data was changed
      if (result.unchanged === 1) {
        return result.changes[0].old_val;
      }

      return null;
    });
};

const deleteChannel = (channelId: string, userId: string): Promise<Boolean> => {
  return db
    .table('channels')
    .get(channelId)
    .update(
      {
        deletedBy: userId,
        deletedAt: new Date(),
        slug: db.uuid(),
      },
      {
        returnChanges: true,
        nonAtomic: true,
      }
    )
    .run();
};

// prettier-ignore
const archiveChannel = (channelId: string): Promise<DBChannel> => {
  return db
    .table('channels')
    .get(channelId)
    .update({ archivedAt: new Date() }, { returnChanges: 'always' })
    .run()
    .then(result => {
      return result.changes[0].new_val || result.changes[0].old_val;
    });
};

// prettier-ignore
const restoreChannel = (channelId: string): Promise<DBChannel> => {
  return db
    .table('channels')
    .get(channelId)
    .update({ archivedAt: db.literal() }, { returnChanges: 'always' })
    .run()
    .then(result => {
      return result.changes[0].new_val || result.changes[0].old_val;
    });
};

// prettier-ignore
const archiveAllPrivateChannels = async (communityId: string) => {
  const channels = await db
    .table('channels')
    .getAll(communityId, { index: 'communityId' })
    .filter({ isPrivate: true })
    .run();

  if (!channels || channels.length === 0) return;

  return await db
    .table('channels')
    .getAll(communityId, { index: 'communityId' })
    .filter({ isPrivate: true })
    .update({ archivedAt: new Date() })
    .run();
};

const incrementMemberCount = (channelId: string): Promise<DBChannel> => {
  return db
    .table('channels')
    .get(channelId)
    .update(
      {
        memberCount: db
          .row('memberCount')
          .default(0)
          .add(1),
      },
      { returnChanges: true }
    )
    .run()
    .then(result => result.changes[0].new_val || result.changes[0].old_val);
};

const decrementMemberCount = (channelId: string): Promise<DBChannel> => {
  return db
    .table('channels')
    .get(channelId)
    .update(
      {
        memberCount: db
          .row('memberCount')
          .default(1)
          .sub(1),
      },
      { returnChanges: true }
    )
    .run()
    .then(result => result.changes[0].new_val || result.changes[0].old_val);
};

const setMemberCount = (
  channelId: string,
  value: number
): Promise<DBChannel> => {
  return db
    .table('channels')
    .get(channelId)
    .update(
      {
        memberCount: value,
      },
      { returnChanges: true }
    )
    .run()
    .then(result => result.changes[0].new_val || result.changes[0].old_val);
};

const getChannelsOnlineMemberCounts = (channelIds: Array<string>) => {
  return db
    .table('usersChannels')
    .getAll(...channelIds, {
      index: 'channelId',
    })
    .filter({ isBlocked: false, isMember: true })
    .pluck(['channelId', 'userId'])
    .eqJoin('userId', db.table('users'))
    .pluck('left', { right: ['lastSeen', 'isOnline'] })
    .zip()
    .filter(rec =>
      rec('isOnline')
        .eq(true)
        .or(
          rec('lastSeen')
            .toEpochTime()
            .ge(
              db
                .now()
                .toEpochTime()
                .sub(86400)
            )
        )
    )
    .group('channelId')
    .count()
    .run();
};

module.exports = {
  getChannelBySlug,
  getChannelById,
  getChannelsByUser,
  getChannelsByCommunity,
  getPublicChannelsByCommunity,
  getChannelsByUserAndCommunity,
  createChannel,
  createGeneralChannel,
  editChannel,
  deleteChannel,
  getChannelsMemberCounts,
  getChannelsThreadCounts,
  getChannels,
  archiveChannel,
  restoreChannel,
  archiveAllPrivateChannels,
  incrementMemberCount,
  decrementMemberCount,
  setMemberCount,
  getChannelsOnlineMemberCounts,
  __forQueryTests: {
    channelsByCommunitiesQuery,
    channelsByIdsQuery,
    threadsByChannelsQuery,
    membersByChannelsQuery,
  },
};
