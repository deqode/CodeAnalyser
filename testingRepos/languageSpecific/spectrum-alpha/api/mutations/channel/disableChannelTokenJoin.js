// @flow
import type { GraphQLContext } from '../../';
import UserError from '../../utils/UserError';
import {
  getOrCreateChannelSettings,
  disableChannelTokenJoin,
} from '../../models/channelSettings';
import {
  isAuthedResolver as requireAuth,
  canModerateChannel,
} from '../../utils/permissions';

type Input = {
  input: {
    id: string,
  },
};

export default requireAuth(async (_: any, args: Input, ctx: GraphQLContext) => {
  const { id: channelId } = args.input;
  const { user, loaders } = ctx;

  if (!(await canModerateChannel(user.id, channelId, loaders))) {
    return new UserError('You donâ€™t have permission to manage this channel');
  }

  return await getOrCreateChannelSettings(channelId).then(
    async () => await disableChannelTokenJoin(channelId)
  );
});
