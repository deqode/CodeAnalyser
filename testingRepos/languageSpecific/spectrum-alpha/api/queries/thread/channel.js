// @flow
import type { GraphQLContext } from '../../';
import type { DBThread } from 'shared/types';

export default async (root: DBThread, _: any, ctx: GraphQLContext) => {
  const { channelId, id } = root;
  const { loaders } = ctx;
  const channel = await loaders.channel.load(channelId);
  if (!channel) {
    console.error(
      'User queried thread of non-existent/deleted channel: ',
      channelId
    );
    console.error('Thread queried: ', id);
  }
  return channel;
};
