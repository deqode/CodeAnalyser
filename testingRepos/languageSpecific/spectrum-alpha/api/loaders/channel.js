// @flow
import {
  getChannels,
  getChannelsThreadCounts,
  getChannelsMemberCounts,
  getChannelsOnlineMemberCounts,
} from '../models/channel';
import { getChannelsSettings } from '../models/channelSettings';
import createLoader from './create-loader';
import { getPendingUsersInChannels } from '../models/usersChannels';

export const __createChannelLoader = createLoader(channels =>
  getChannels(channels)
);

export const __createChannelThreadCountLoader = createLoader(
  channels => getChannelsThreadCounts(channels),
  'group'
);

export const __createChannelPendingMembersLoader = createLoader(
  channels => getPendingUsersInChannels(channels),
  'group'
);

export const __createChannelMemberCountLoader = createLoader(
  channels => getChannelsMemberCounts(channels),
  'group'
);

export const __createChannelOnlineMemberCountLoader = createLoader(
  channelIds => getChannelsOnlineMemberCounts(channelIds),
  'group'
);

export const __createChannelSettingsLoader = createLoader(
  channelIds => getChannelsSettings(channelIds),
  key => key.channelId
);

export default () => {
  throw new Error(
    '⚠️ Do not import loaders directly, get them from the GraphQL context instead! ⚠️'
  );
};
