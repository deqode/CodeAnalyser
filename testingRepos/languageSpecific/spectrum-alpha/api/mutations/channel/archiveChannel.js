// @flow
import type { GraphQLContext } from '../../';
import UserError from '../../utils/UserError';
import { archiveChannel } from '../../models/channel';
import {
  isAuthedResolver as requireAuth,
  canModerateChannel,
} from '../../utils/permissions';

type Input = {
  input: {
    channelId: string,
  },
};

export default requireAuth(async (_: any, args: Input, ctx: GraphQLContext) => {
  const { channelId } = args.input;
  const { user, loaders } = ctx;

  const channelToEvaluate = await loaders.channel.load(channelId);

  if (!(await canModerateChannel(user.id, channelId, loaders))) {
    return new UserError('You don’t have permission to archive this channel');
  }

  if (channelToEvaluate.archivedAt) {
    return new UserError('This channel is already archived');
  }

  if (channelToEvaluate.slug === 'general') {
    return new UserError(
      'The general channel in a community can’t be archived'
    );
  }

  return await archiveChannel(channelId);
});
