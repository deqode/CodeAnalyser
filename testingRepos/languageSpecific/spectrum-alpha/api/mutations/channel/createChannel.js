// @flow
import type { GraphQLContext } from '../../';
import type { CreateChannelInput } from '../../models/channel';
import UserError from '../../utils/UserError';
import { channelSlugIsDenyListed } from '../../utils/permissions';
import { getChannelBySlug, createChannel } from '../../models/channel';
import { createOwnerInChannel } from '../../models/usersChannels';
import {
  isAuthedResolver as requireAuth,
  canModerateCommunity,
} from '../../utils/permissions';

export default requireAuth(
  async (_: any, args: CreateChannelInput, ctx: GraphQLContext) => {
    const { user, loaders } = ctx;

    // TODO: Figure out how to not have to do this - somehow combine forces with canModerateChannel function which is fetching most of the same data anyways
    const community = await loaders.community.load(args.input.communityId);

    if (
      !(await canModerateCommunity(user.id, args.input.communityId, loaders))
    ) {
      return new UserError(
        'You donâ€™t have permission to create channels in this community'
      );
    }

    if (channelSlugIsDenyListed(args.input.slug)) {
      return new UserError(
        'This channel url is reserved - please try another name'
      );
    }

    const channelWithSlug = await getChannelBySlug(
      args.input.slug,
      community.slug
    );

    if (channelWithSlug) {
      return new UserError(
        'A channel with this url already exists in this community - please try another name'
      );
    }

    const newChannel = await createChannel(args, user.id);

    return await createOwnerInChannel(newChannel.id, user.id).then(
      () => newChannel
    );
  }
);
