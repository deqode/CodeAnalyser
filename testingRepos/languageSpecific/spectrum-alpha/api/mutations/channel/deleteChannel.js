// @flow
import type { GraphQLContext } from '../../';
import UserError from '../../utils/UserError';
import { removeMembersInChannel } from '../../models/usersChannels';
import { getChannelById, deleteChannel } from '../../models/channel';
import { getThreadsByChannelToDelete, deleteThread } from '../../models/thread';
import {
  isAuthedResolver as requireAuth,
  canModerateChannel,
} from '../../utils/permissions';

type Input = {
  channelId: string,
};

export default requireAuth(async (_: any, args: Input, ctx: GraphQLContext) => {
  const { channelId } = args;
  const { user, loaders } = ctx;

  if (!(await canModerateChannel(user.id, channelId, loaders))) {
    return new UserError('You donâ€™t have permission to manage this channel');
  }

  const channel = await getChannelById(channelId);

  if (channel.slug === 'general') {
    return new UserError("The general channel can't be deleted");
  }

  const [allThreadsInChannel] = await Promise.all([
    getThreadsByChannelToDelete(channelId),
    deleteChannel(channelId, user.id),
    removeMembersInChannel(channelId),
  ]);

  if (allThreadsInChannel.length === 0) return true;

  return allThreadsInChannel.map(
    async thread => await deleteThread(thread.id, user.id)
  );
});
