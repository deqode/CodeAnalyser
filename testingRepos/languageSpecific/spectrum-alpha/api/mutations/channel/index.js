// @flow
import createChannel from './createChannel';
import deleteChannel from './deleteChannel';
import editChannel from './editChannel';
import toggleChannelSubscription from './toggleChannelSubscription';
import toggleChannelNotifications from './toggleChannelNotifications';
import togglePendingUser from './togglePendingUser';
import unblockUser from './unblockUser';
import archiveChannel from './archiveChannel';
import restoreChannel from './restoreChannel';
import joinChannelWithToken from './joinChannelWithToken';
import enableChannelTokenJoin from './enableChannelTokenJoin';
import disableChannelTokenJoin from './disableChannelTokenJoin';
import resetChannelJoinToken from './resetChannelJoinToken';
import updateChannelSlackBotLinks from './updateChannelSlackBotLinks';

module.exports = {
  Mutation: {
    createChannel,
    deleteChannel,
    editChannel,
    toggleChannelSubscription,
    toggleChannelNotifications,
    togglePendingUser,
    unblockUser,
    archiveChannel,
    restoreChannel,
    joinChannelWithToken,
    enableChannelTokenJoin,
    disableChannelTokenJoin,
    resetChannelJoinToken,
    updateChannelSlackBotLinks,
  },
};
