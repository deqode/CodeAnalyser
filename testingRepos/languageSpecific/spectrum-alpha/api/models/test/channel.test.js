// @flow

import * as channel from '../channel';

const queries = channel.__forQueryTests;

import {
  BRIAN_ID,
  MAX_ID,
  SPECTRUM_COMMUNITY_ID,
  SPECTRUM_GENERAL_CHANNEL_ID,
  SPECTRUM_PRIVATE_CHANNEL_ID,
  DELETED_COMMUNITY_DELETED_CHANNEL_ID,
} from 'api/migrations/seed/default/constants';

describe('models/channel', () => {
  describe('channelsByCommunitiesQuery', () => {
    it('excludes deleted channels', async () => {
      const channels = await queries
        .channelsByCommunitiesQuery(SPECTRUM_COMMUNITY_ID)
        .run();
      expect(
        channels.filter(channel => channel.deletedAt !== undefined)
      ).toEqual([]);
    });
  });

  describe('channelsByIdsQuery', () => {
    it('excludes deleted channels', async () => {
      const channels = await queries
        .channelsByIdsQuery(
          SPECTRUM_GENERAL_CHANNEL_ID,
          DELETED_COMMUNITY_DELETED_CHANNEL_ID
        )
        .run();
      expect(
        channels.filter(channel => channel.deletedAt !== undefined)
      ).toEqual([]);
    });
  });

  describe('threadsByChannelsQuery', () => {
    it('excludes deleted channels', async () => {
      const threads = await queries
        .threadsByChannelsQuery(DELETED_COMMUNITY_DELETED_CHANNEL_ID)
        .run();
      expect(threads).toEqual([]);
    });

    it('excludes deleted threads', async () => {
      const threads = await queries
        .threadsByChannelsQuery(SPECTRUM_GENERAL_CHANNEL_ID)
        .run();
      expect(threads.filter(thread => thread.deletedAt !== undefined)).toEqual(
        []
      );
    });
  });

  describe('membersByChannelsQuery', () => {
    it('excludes deleted channels', async () => {
      const members = await queries
        .membersByChannelsQuery(DELETED_COMMUNITY_DELETED_CHANNEL_ID)
        .run();
      expect(members).toEqual([]);
    });

    it('excludes removed members', async () => {
      const members = await queries
        .membersByChannelsQuery(SPECTRUM_GENERAL_CHANNEL_ID)
        .run();
      expect(members.filter(member => !member.isMember)).toEqual([]);
    });
  });

  describe('getChannelsByCommunity', () => {
    it('returns correct set of channels', async () => {
      expect(
        await channel.getChannelsByCommunity(SPECTRUM_COMMUNITY_ID)
      ).toMatchSnapshot();
    });
  });

  describe('getPublicChannelsByCommunity', () => {
    it('returns correct set of channels', async () => {
      expect(
        await channel.getPublicChannelsByCommunity(SPECTRUM_COMMUNITY_ID)
      ).toMatchSnapshot();
    });
  });

  describe('getChannelsByUserAndCommunity', () => {
    it('returns correct set of channels', async () => {
      expect(
        await channel.getChannelsByUserAndCommunity(
          SPECTRUM_COMMUNITY_ID,
          MAX_ID
        )
      ).toMatchSnapshot();
    });
  });

  describe('getChannelsByUser', () => {
    it('returns correct set of channels', async () => {
      expect(await channel.getChannelsByUser(BRIAN_ID)).toMatchSnapshot();
    });
  });

  describe('getChannelBySlug', () => {
    it('excludes deleted channels', async () => {
      expect(await channel.getChannelBySlug('deleted', 'spectrum')).toEqual(
        null
      );
    });
  });

  describe('getChannelById', () => {
    it('excludes deleted channels', async () => {
      expect(
        await channel.getChannelById(DELETED_COMMUNITY_DELETED_CHANNEL_ID)
      ).toEqual(null);
    });
  });

  describe('getChannels', () => {
    it('excludes deleted channels', async () => {
      expect(
        await channel.getChannels([
          SPECTRUM_GENERAL_CHANNEL_ID,
          DELETED_COMMUNITY_DELETED_CHANNEL_ID,
        ])
      ).toMatchSnapshot();
    });
  });

  describe('getChannelsThreadCounts', () => {
    it('excludes deleted channels', async () => {
      expect(
        await channel.getChannelsThreadCounts([
          SPECTRUM_GENERAL_CHANNEL_ID,
          DELETED_COMMUNITY_DELETED_CHANNEL_ID,
        ])
      ).toMatchSnapshot();
    });

    it('excludes deleted threads', async () => {
      expect(
        await channel.getChannelsThreadCounts([
          SPECTRUM_GENERAL_CHANNEL_ID,
          SPECTRUM_PRIVATE_CHANNEL_ID,
        ])
      ).toMatchSnapshot();
    });
  });

  describe('getChannelsMemberCounts', () => {
    it('excludes deleted channels', async () => {
      expect(
        await channel.getChannelsMemberCounts([
          SPECTRUM_GENERAL_CHANNEL_ID,
          DELETED_COMMUNITY_DELETED_CHANNEL_ID,
        ])
      ).toMatchSnapshot();
    });

    it('excludes non-members', async () => {
      expect(
        await channel.getChannelsMemberCounts([
          SPECTRUM_GENERAL_CHANNEL_ID,
          SPECTRUM_PRIVATE_CHANNEL_ID,
        ])
      ).toMatchSnapshot();
    });
  });
});
