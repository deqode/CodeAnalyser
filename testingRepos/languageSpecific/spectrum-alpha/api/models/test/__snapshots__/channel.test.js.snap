// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models/channel getChannels excludes deleted channels 1`] = `
Array [
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "General chatter",
    "id": "1",
    "isDefault": true,
    "isPrivate": false,
    "memberCount": 5,
    "name": "General",
    "slug": "general",
  },
]
`;

exports[`models/channel getChannelsByCommunity returns correct set of channels 1`] = `
Array [
  Object {
    "archivedAt": 2016-12-31T23:00:00.000Z,
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Testing archiving",
    "id": "5",
    "isDefault": true,
    "isPrivate": false,
    "memberCount": 3,
    "name": "Archived",
    "slug": "archived",
  },
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "General chatter",
    "id": "1",
    "isDefault": true,
    "isPrivate": false,
    "memberCount": 5,
    "name": "General",
    "slug": "general",
  },
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Private chatter",
    "id": "2",
    "isDefault": false,
    "isPrivate": true,
    "memberCount": 5,
    "name": "Private",
    "slug": "private",
  },
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Moderator created channel",
    "id": "8",
    "isDefault": false,
    "isPrivate": false,
    "memberCount": 1,
    "name": "Moderator created",
    "slug": "moderator-created",
  },
]
`;

exports[`models/channel getChannelsByUser returns correct set of channels 1`] = `
Array [
  Object {
    "communityId": "2",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Private chatter",
    "id": "4",
    "isBlocked": false,
    "isDefault": false,
    "isMember": true,
    "isModerator": false,
    "isOwner": false,
    "isPending": false,
    "isPrivate": true,
    "memberCount": 5,
    "name": "Private",
    "receiveNotifications": true,
    "slug": "private",
  },
  Object {
    "communityId": "5",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "General",
    "id": "11",
    "isBlocked": false,
    "isDefault": false,
    "isMember": true,
    "isModerator": false,
    "isOwner": false,
    "isPending": false,
    "isPrivate": false,
    "memberCount": 1,
    "name": "General",
    "receiveNotifications": false,
    "slug": "general",
  },
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Private chatter",
    "id": "2",
    "isBlocked": false,
    "isDefault": false,
    "isMember": true,
    "isModerator": false,
    "isOwner": false,
    "isPending": false,
    "isPrivate": true,
    "memberCount": 5,
    "name": "Private",
    "receiveNotifications": true,
    "slug": "private",
  },
  Object {
    "communityId": "1",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "General chatter",
    "id": "1",
    "isBlocked": false,
    "isDefault": true,
    "isMember": true,
    "isModerator": false,
    "isOwner": false,
    "isPending": false,
    "isPrivate": false,
    "memberCount": 5,
    "name": "General",
    "receiveNotifications": true,
    "slug": "general",
  },
  Object {
    "communityId": "2",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "Payments Features",
    "id": "10",
    "isBlocked": false,
    "isDefault": false,
    "isMember": true,
    "isModerator": false,
    "isOwner": true,
    "isPending": false,
    "isPrivate": false,
    "memberCount": 5,
    "name": "Payments Features",
    "receiveNotifications": true,
    "slug": "features",
  },
  Object {
    "communityId": "2",
    "createdAt": 2016-12-31T23:00:00.000Z,
    "description": "General chatter",
    "id": "3",
    "isBlocked": false,
    "isDefault": true,
    "isMember": true,
    "isModerator": false,
    "isOwner": false,
    "isPending": false,
    "isPrivate": false,
    "memberCount": 5,
    "name": "General",
    "receiveNotifications": true,
    "slug": "general",
  },
]
`;

exports[`models/channel getChannelsByUserAndCommunity returns correct set of channels 1`] = `
Array [
  "2",
  "1",
]
`;

exports[`models/channel getChannelsMemberCounts excludes deleted channels 1`] = `
Array [
  Object {
    "group": "1",
    "reduction": 5,
  },
]
`;

exports[`models/channel getChannelsMemberCounts excludes non-members 1`] = `
Array [
  Object {
    "group": "1",
    "reduction": 5,
  },
  Object {
    "group": "2",
    "reduction": 5,
  },
]
`;

exports[`models/channel getChannelsThreadCounts excludes deleted channels 1`] = `
Array [
  Object {
    "group": "1",
    "reduction": 4,
  },
]
`;

exports[`models/channel getChannelsThreadCounts excludes deleted threads 1`] = `
Array [
  Object {
    "group": "1",
    "reduction": 4,
  },
  Object {
    "group": "2",
    "reduction": 3,
  },
]
`;

exports[`models/channel getPublicChannelsByCommunity returns correct set of channels 1`] = `
Array [
  "1",
  "8",
]
`;
